Support substring and replacement #8

pombredanne opened the pull request:
This PR adds minimal support for some Bash idioms for #8

- simple replacement as in ${foo/bar/baz}
- substrings as in  ${foo:4:2}

In addition:
- add support for plain $parameter substitution including
some simple nesting.
- apply some minor code cleanups

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Add support for simple bash substring expansion

"Substring Expansion" is a bash'ism to extract slices from a string
in the form of:
 ${parameter:offset}
 ${parameter:offset:length}

This is a minimal implementation that only supports simple slices on
strings and does not support arrays

See https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Add support for simple bash replacement expansion

Replacing with "/" is a bash'ism to replace a pattern with a string in
the parameter being expanded with the form of:
 ${parameter/pattern/string}

This is a simple and minimal implementation:
- supports only plain strings patterns (and not actual glob patterns).
- does not handle #, %, @ and * pattern modifiers.

See https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Import os module

Using expand(env=None) was failing as the os module was not imported

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Add new pe.expand_simple() function

This performs a simple expansion for plain $parameters (without curly
braces) and is done as a pre-processing step in expand(). This also
allows for some minimal level of parameter nesting as a $parameter can
be nested in a ${parameter}.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Remove unused variable

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pombredanne committed:
Enable more simple nested expansions

Also update module documentation

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve docstrings

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne commented:
Note that I have tried to keep each commit mostly self-standing if you want to cherry pick some

kojiromike commented on parameter_expansion/pe.py:
```diff
@@ -130,8 +151,53 @@ def follow_brace(shl, env):
+                replace_all = True
+                arg1 = next(shl)
+
+            sep = next(shl)
```
Leaving off the final slash is valid.

```bash
$ parameter=aa/bb/cc; echo "-${parameter/aa}-"
-/bb/cc-
```

I think you can implement this by the default argument to `next`: `sep = next(shl, "/")`

pombredanne replied:
Done. Thanks

kojiromike commented on parameter_expansion/tests/test_pe.py:
```diff
@@ -79,6 +79,78 @@
+        "aa/bb/cc",
+        "-bb/bb/cc-",
+    ),
+    "-${parameter/aa/}-": (
```
If you implement

```bash
$ parameter=aa/bb/cc; echo "-${parameter/aa}-"
-/bb/cc-
```

please add a test case for it.

pombredanne replied:
Done. Thanks

pombredanne replied:
I also added a case for:
```
$ parameter=aa/bb/aa; echo "-${parameter//aa}-"
-/bb/-
```

kojiromike commented on parameter_expansion/tests/test_pe.py:
```diff
@@ -79,6 +79,78 @@
+simple_test_cases = {
+    # string,               env, result
+    "-${parameter/$aa/$bb}-": (
+        dict(
```
Why `dict()` instead of a literal dict?

pombredanne replied:
I feel it makes it stand out more clearly as being a mapping of environment variables... but I am fine either way

kojiromike reviewed (approved):
Very nice, thanks. I am approving this as-is. You can implement my additional suggestions, or just merge as-is.

pombredanne committed:
Allow no trailing slash in substitutions

Leaving off the final slash is valid in a substitution as in:
$ parameter=aa/bb/cc; echo "-${parameter/aa}-"
-/bb/cc-

Reported-by: Michael A. Smith <michael@smith-li.com>
Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Merge remote-tracking branch 'origin/main' into bashisms

pombredanne commented:
Thank you for the swift merge.
